src/content/language/macros.md

```diff
@@ -0,0 +1,287 @@
+and operates on it during compile time to create, modify, or add declarations.
+
+You can think about the macro system in two parts: using macros and writing macros.
+This page covers each (at a high level, as ***the feature is still under development***)
+in the following sections:
+
+- [**The `JsonCodable` macro**](#the-jsoncodable-macro):
+A ready-made macro you can try out today (behind an experiment flag)
+that offers a seamless solution to the
+common issue of tedious JSON serialization in Dart.
+
+- [**The macros feature in general**](#the-macros-language-feature):
+Why we're adding macros to Dart, motivating use cases,
+benefits over existing code gen solutions,
+and a cursory overview of how writing macros works.
+
+[spec]: {{site.repo.dart.lang}}/blob/main/working/macros/feature-specification.md
+[motivation]: {{site.repo.dart.lang}}/language/blob/main/working/macros/motivation.md
+
+## The `JsonCodable` macro
+
+:::important
+The `JsonCodable` macro is not stable and currently behind an [experiment flag][].
+Functionality is subject to change.
+:::
+
+The [`JsonCodable`][] macro encodes and decodes
+user-defined Dart classes to JSON maps of type `Map<String, Object?>`.
+It generates two members, a `toJson` serialization method,
+and a `fromJson` deserialization constructor.
+
+[experiment flag]: /tools/experiment-flags
+[`JsonCodable`]: {{site.repo.dart.sdk}}/tree/main/pkg/json
+
+### Set up the experiment
+
+1. [Add the package][] to your pubspec and retrieve
+  its dependencies.
+2. [Add the experiment][] to the `analysis_options.yaml`
+  file at the root of your project:
+
+   ```yaml
+   analyzer:
+    enable-experiment:
+      - macros
+   ```
+3. Import it in the file you plan to use it:
+
+  ```dart
+  import 'package:json/json.dart';
+  ```
+
+4. Run your project with the experiment flag:
+
+  ```console
+  dart --enable-experiment=macros run bin/my_app.dart
+  ```
+
+[Add the package]: /guides/packages
+[Add the experiment]: /tools/experiment-flags#using-experiment-flags-with-the-dart-analyzer-command-line-and-ide
+
+### Use the macro
+
+To use the `JsonCodable` macro, append the annotation to the class you want to serialize:
+
+```dart
+import 'package:json/json.dart';
+
+@JsonCodable() // Macro annotation.
+class User {
+ final int? age;
+ final String name;
+ final String username;
+}
+```
+
+The macro introspects the `User` class and derives the implementations of
+`fromJson` and `toJson` using the `User` class's fields.
+
+So, without needing to define them yourself, `toJson` and `fromJson` are now
+available to use on objects of the annotated class:
+
+```dart
+void main() {
+ // Given some arbitrary JSON:
+ var userJson = {
+   'age': 5,
+   'name': 'Roger',
+   'username': 'roger1337'
+ };
+
+ // Use the generated members:
+ var user = User.fromJson(userJson);
+ print(user);
+ print(user.toJson());
+}
+```
+
+### View the generated code
+
+You can optionally view the generated code.
```

Comment:
Maybe lead with something like "Sometimes it can be useful to view the generated code to better understand how a macros works, or to inspect the details of what it offers". ?

Reply:
Should we mention other reasons like "if you have a stack trace in the generated code and want to be able to look at it, or if you need to be able to set break points to inspect"? Or keep it high-level/as-is?